File: C++/7-Trees/Problems/3-Diameter-of-BinaryTree/Diameter-of-BinaryTree.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>
#include <stack>
#include <climits>
#include <functional> // Include for std::function
#include <unordered_map> // Include for std::unordered_map

using namespace std;

// Definition of a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// Function to calculate the diameter of a binary tree.

// Approach 1: Recursive (Naive)
// Time Complexity: O(n^2) in the worst case (skewed tree), O(n log n) for balanced tree
// Space Complexity: O(height) -  due to recursive call stack,  O(n) in worst case (skewed tree)
int diameterOfBinaryTree_Recursive(TreeNode* root) {
    if (!root) {
        return 0;
    }

    // Calculate the height of the left and right subtrees.
    int leftHeight = (root->left) ? 1 + max(diameterOfBinaryTree_Recursive(root->left->left), diameterOfBinaryTree_Recursive(root->left->right)) : 0;
    int rightHeight = (root->right) ? 1 + max(diameterOfBinaryTree_Recursive(root->right->left), diameterOfBinaryTree_Recursive(root->right->right)) : 0;

    // Diameter through the root.
    int diameterThroughRoot = leftHeight + rightHeight;

    // Recursively calculate the diameter of the left and right subtrees.
    int leftDiameter = diameterOfBinaryTree_Recursive(root->left);
    int rightDiameter = diameterOfBinaryTree_Recursive(root->right);

    // Return the maximum of the three diameters.
    return max({diameterThroughRoot, leftDiameter, rightDiameter});
}

// Approach 2: Recursive (Optimized)
// Time Complexity: O(n) - Each node is visited exactly once.
// Space Complexity: O(height) - due to recursive call stack, O(n) in worst case.
int diameterOfBinaryTree_Optimized(TreeNode* root, int* diameter) {
    if (!root) {
        return 0;
    }

    // Recursively calculate the height of the left and right subtrees.
    int leftHeight = diameterOfBinaryTree_Optimized(root->left, diameter);
    int rightHeight = diameterOfBinaryTree_Optimized(root->right, diameter);

    // Calculate the diameter through the current root.
    *diameter = max(*diameter, leftHeight + rightHeight);

    // Return the height of the current node.
    return 1 + max(leftHeight, rightHeight);
}

int diameterOfBinaryTree_Optimized(TreeNode* root) {
    int diameter = 0;
    diameterOfBinaryTree_Optimized(root, &diameter);
    return diameter;
}


// Approach 3: Iterative using Level Order Traversal (BFS) - Not very efficient for diameter, more complex.
// Time Complexity: O(n^2) in worst case, where calculating height of subtrees can take O(n) for each node.
// Space Complexity: O(w) - where w is the maximum width of the tree.  O(n) in worst case.
int diameterOfBinaryTree_BFS(TreeNode* root) {
    if (!root) return 0;

    int diameter = 0;
    queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        TreeNode* node = q.front();
        q.pop();

        // Helper function to calculate height (can be optimized, but kept here for clarity within BFS)
        function<int(TreeNode*)> calculateHeight = [&](TreeNode* node) {
            if (!node) return 0;
            return 1 + max(calculateHeight(node->left), calculateHeight(node->right));
        };

        int leftHeight = (node->left) ? calculateHeight(node->left) : 0;
        int rightHeight = (node->right) ? calculateHeight(node->right) : 0;
        diameter = max(diameter, leftHeight + rightHeight);

        if (node->left) q.push(node->left);
        if (node->right) q.push(node->right);
    }
    return diameter;
}

// Approach 4: Iterative using Stack (Pre-order Traversal) -  More complex, not ideal for diameter
// Time Complexity: O(n) - Each node is visited.  Height calculations within the loop make it less efficient for diameter.
// Space Complexity: O(h) - max height of the tree. O(n) in worst case.

int diameterOfBinaryTree_Iterative(TreeNode* root) {
    if (!root) return 0;

    int diameter = 0;
    stack<TreeNode*> s;
    s.push(root);

    // Helper function to calculate height
    function<int(TreeNode*)> calculateHeight = [&](TreeNode* node) {
        if (!node) return 0;
        return 1 + max(calculateHeight(node->left), calculateHeight(node->right));
    };
    while (!s.empty()) {
        TreeNode* node = s.top();
        s.pop();

        int leftHeight = (node->left) ? calculateHeight(node->left) : 0;
        int rightHeight = (node->right) ? calculateHeight(node->right) : 0;
        diameter = max(diameter, leftHeight + rightHeight);

        if (node->right) s.push(node->right); // Push right child first
        if (node->left) s.push(node->left);    // Push left child
    }
    return diameter;
}

// Approach 5:  Single Pass Post-Order Traversal (Most Efficient Iterative)
// This approach calculates the height and diameter in a single pass, iteratively.
// It mimics the optimized recursive approach, but uses a stack to avoid recursion.
// Time Complexity: O(n) - Each node is visited exactly once.
// Space Complexity: O(h) - Maximum height of the tree.  O(n) in the worst case (skewed tree).

int diameterOfBinaryTree_IterativePostOrder(TreeNode* root) {
    if (!root) return 0;

    int diameter = 0;
    stack<pair<TreeNode*, int>> stack; // Store node and its state (0: visit children, 1: process node)
    unordered_map<TreeNode*, int> heightMap; // Store calculated heights of nodes
    stack.push({root, 0});

    while (!stack.empty()) {
        pair<TreeNode*, int> current = stack.top();
        TreeNode* node = current.first;
        int state = current.second;
        stack.pop();

        if (state == 0) { // Visiting children
            stack.push({node, 1}); // Mark node to be processed after children
            if (node->right) stack.push({node->right, 0}); // Push right child first for post-order
            if (node->left) stack.push({node->left, 0});   // Then push left child
        } else { // Processing node
            int leftHeight = (node->left) ? heightMap[node->left] : 0;
            int rightHeight = (node->right) ? heightMap[node->right] : 0;

            // Calculate diameter through this node
            diameter = max(diameter, leftHeight + rightHeight);

            // Calculate height of this node and store it.
            int nodeHeight = 1 + max(leftHeight, rightHeight);
            heightMap[node] = nodeHeight;
        }
    }
    return diameter;
}


// Helper function to create a binary tree (for testing)
TreeNode* createBinaryTree(const vector<int>& values, int index) {
    if (index >= values.size() || values[index] == -1) { // -1 represents null
        return nullptr;
    }
    TreeNode* node = new TreeNode(values[index]);
    node->left = createBinaryTree(values, 2 * index + 1);
    node->right = createBinaryTree(values, 2 * index + 2);
    return node;
}

// Helper function to print the tree (for visualization) -  Using Level Order
void printBinaryTree(TreeNode* root) {
    if (!root) {
        cout << "Tree is empty" << endl;
        return;
    }

    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty()) {
        int levelSize = q.size();
        for (int i = 0; i < levelSize; ++i) {
            TreeNode* node = q.front();
            q.pop();
            if (node) {
                cout << node->val << " ";
                q.push(node->left);
                q.push(node->right);
            } else {
                cout << "null ";
            }
        }
        cout << endl;
    }
}

int main() {
    // Example usage:
    // Create a binary tree: [1,2,3,4,5,null,null,null,null,null,null]  (see printout below)
    vector<int> treeNodes = {1, 2, 3, 4, 5, -1, -1, -1, -1, -1, -1}; // -1 represents null
    TreeNode* root = createBinaryTree(treeNodes, 0);

    cout << "Binary Tree (Level Order):" << endl;
    printBinaryTree(root);

    cout << "\nDiameter of Binary Tree:" << endl;
    cout << "Approach 1 (Recursive): " << diameterOfBinaryTree_Recursive(root) << endl;
    cout << "Approach 2 (Optimized Recursive): " << diameterOfBinaryTree_Optimized(root) << endl;
    cout << "Approach 3 (BFS): " << diameterOfBinaryTree_BFS(root) << endl;
    cout << "Approach 4 (Iterative): " << diameterOfBinaryTree_Iterative(root) << endl;
    cout << "Approach 5 (Iterative Post-Order): " << diameterOfBinaryTree_IterativePostOrder(root) << endl;


    // Example 2:
    vector<int> treeNodes2 = {1, 2, 3, 4, 5, 6, 7};
    TreeNode* root2 = createBinaryTree(treeNodes2, 0);
    cout << "\nBinary Tree 2 (Level Order):" << endl;
    printBinaryTree(root2);
    cout << "\nDiameter of Binary Tree 2:" << endl;
    cout << "Approach 1 (Recursive): " << diameterOfBinaryTree_Recursive(root2) << endl;
    cout << "Approach 2 (Optimized Recursive): " << diameterOfBinaryTree_Optimized(root2) << endl;
    cout << "Approach 3 (BFS): " << diameterOfBinaryTree_BFS(root2) << endl;
    cout << "Approach 4 (Iterative): " << diameterOfBinaryTree_Iterative(root2) << endl;
    cout << "Approach 5 (Iterative Post-Order): " << diameterOfBinaryTree_IterativePostOrder(root2) << endl;

    // Example 3: Skewed Tree (Worst Case for some approaches)
    vector<int> treeNodes3 = {1, 2, -1, 3, -1, -1, -1, 4, -1, -1, -1, -1, -1, -1, -1, 5}; // Skewed to the left
    TreeNode* root3 = createBinaryTree(treeNodes3, 0);
    cout << "\nBinary Tree 3 (Skewed - Level Order):" << endl;
    printBinaryTree(root3);
    cout << "\nDiameter of Binary Tree 3 (Skewed):" << endl;
    cout << "Approach 1 (Recursive): " << diameterOfBinaryTree_Recursive(root3) << endl;
    cout << "Approach 2 (Optimized Recursive): " << diameterOfBinaryTree_Optimized(root3) << endl;
    cout << "Approach 3 (BFS): " << diameterOfBinaryTree_BFS(root3) << endl;
    cout << "Approach 4 (Iterative): " << diameterOfBinaryTree_Iterative(root3) << endl;
    cout << "Approach 5 (Iterative Post-Order): " << diameterOfBinaryTree_IterativePostOrder(root3) << endl;


    return 0;
}

File: Java/7-Trees/Problems/3-Diameter-of-BinaryTree/Diameter-of-BinaryTree.java
import java.util.*;
import java.lang.*;

class Solution {

    /**
     * Approach 1: Recursive with Global Variable
     * -   Calculates the diameter by recursively finding the height of left and right subtrees.
     * -   Updates a global variable `diameter` whenever a larger diameter is found.
     * -   Time Complexity: O(N), where N is the number of nodes.
     * -   Space Complexity: O(H), where H is the height of the tree (call stack).  In the worst case, H = N for a skewed tree.
     */
    int diameter = 0; // Global variable to store the diameter

    public int diameterOfBinaryTree_approach1(TreeNode root) {
        if (root == null) return 0;
        diameter = 0; // Reset diameter
        height(root);
        return diameter;
    }

    private int height(TreeNode node) {
        if (node == null) return 0;

        int leftHeight = height(node.left);
        int rightHeight = height(node.right);

        // Calculate diameter at each node and update the global variable if needed.
        diameter = Math.max(diameter, leftHeight + rightHeight);

        return Math.max(leftHeight, rightHeight) + 1;
    }



    /**
     * Approach 2: Recursive without Global Variable
     * -   Similar to Approach 1, but avoids using a global variable.
     * -   Instead, the diameter is calculated and returned directly from the recursive function.
     * -   Uses a helper function that returns an array: [height, diameter].
     * -   Time Complexity: O(N)
     * -   Space Complexity: O(H)
     */
    public int diameterOfBinaryTree_approach2(TreeNode root) {
        if (root == null) return 0;
        int[] result = heightAndDiameter(root);
        return result[1];
    }

    private int[] heightAndDiameter(TreeNode node) {
        if (node == null) {
            return new int[] {0, 0}; // {height, diameter}
        }

        int[] leftResult = heightAndDiameter(node.left);
        int[] rightResult = heightAndDiameter(node.right);

        int currentHeight = Math.max(leftResult[0], rightResult[0]) + 1;
        int currentDiameter = Math.max(leftResult[0] + rightResult[0], Math.max(leftResult[1], rightResult[1]));

        return new int[] {currentHeight, currentDiameter};
    }



    /**
     * Approach 3: Iterative using Stack (Post-order Traversal)
     * -   Uses a stack to perform a post-order traversal of the tree.
     * -   A HashMap stores the height of each node.
     * -   Calculates the diameter while traversing the tree.
     * -   Time Complexity: O(N)
     * -   Space Complexity: O(N) (for stack and HashMap)
     */
    public int diameterOfBinaryTree_approach3(TreeNode root) {
        if (root == null) return 0;

        Stack<TreeNode> stack = new Stack<>();
        HashMap<TreeNode, Integer> heightMap = new HashMap<>();
        int diameter = 0;
        TreeNode current = root;

        while (current != null || !stack.isEmpty()) {
            if (current != null) {
                stack.push(current);
                current = current.left;
            } else {
                TreeNode top = stack.peek();
                if (top.right != null && heightMap.get(top.right) == null) {
                    current = top.right;
                } else {
                    stack.pop();
                    int leftHeight = (top.left != null) ? heightMap.get(top.left) : 0;
                    int rightHeight = (top.right != null) ? heightMap.get(top.right) : 0;
                    diameter = Math.max(diameter, leftHeight + rightHeight);
                    heightMap.put(top, Math.max(leftHeight, rightHeight) + 1);
                }
            }
        }
        return diameter;
    }

    /**
     * Approach 4: Level Order Traversal (BFS)
     * - Uses Breadth-First Search (BFS) to traverse the tree level by level.
     * - For each node, calculates its height, and the diameter passing through that node.
     * - Time Complexity: O(N^2) in worst case.  For each node, we calculate the height, which can take O(N) in a skewed tree.
     * - Space Complexity: O(W) where W is the maximum width of the tree.  O(N) in the worst case (full binary tree).
     */
    public int diameterOfBinaryTree_approach4(TreeNode root) {
        if (root == null) return 0;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int diameter = 0;

        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            int leftHeight = calculateHeight(node.left);
            int rightHeight = calculateHeight(node.right);
            diameter = Math.max(diameter, leftHeight + rightHeight);

            if (node.left != null) queue.offer(node.left);
            if (node.right != null) queue.offer(node.right);
        }
        return diameter;
    }

    private int calculateHeight(TreeNode node) {
        if (node == null) return 0;
        return Math.max(calculateHeight(node.left), calculateHeight(node.right)) + 1;
    }


    /**
     * Approach 5: Divide and Conquer (Optimized Recursion)
     * -   This is a refined recursive approach that efficiently calculates both height and diameter in a single pass.
     * -   It avoids redundant height calculations by returning both values from each recursive call.
     * -   Time Complexity: O(N), where N is the number of nodes.
     * -   Space Complexity: O(H), where H is the height of the tree (call stack).
     */
    public int diameterOfBinaryTree_approach5(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return postOrder(root)[1];
    }

    private int[] postOrder(TreeNode node) {
        if (node == null) {
            return new int[]{0, 0}; // {height, diameter}
        }

        int[] left = postOrder(node.left);
        int[] right = postOrder(node.right);

        int height = Math.max(left[0], right[0]) + 1;
        int diameter = Math.max(left[0] + right[0], Math.max(left[1], right[1]));

        return new int[]{height, diameter};
    }


    public static void main(String[] args) {
        Solution solution = new Solution();

        // Example usage:
        // Create a sample binary tree:
        //      1
        //     / \
        //    2   3
        //   / \
        //  4   5
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        // Approach 1
        System.out.println("Approach 1: Diameter of Binary Tree: " + solution.diameterOfBinaryTree_approach1(root)); // Expected Output: 3

        // Approach 2
        System.out.println("Approach 2: Diameter of Binary Tree: " + solution.diameterOfBinaryTree_approach2(root)); // Expected Output: 3

        // Approach 3
        System.out.println("Approach 3: Diameter of Binary Tree: " + solution.diameterOfBinaryTree_approach3(root)); // Expected Output: 3

        // Approach 4
        System.out.println("Approach 4: Diameter of Binary Tree: " + solution.diameterOfBinaryTree_approach4(root)); // Expected Output: 3

        // Approach 5
        System.out.println("Approach 5: Diameter of Binary Tree: " + solution.diameterOfBinaryTree_approach5(root)); // Expected Output: 3
    }
}

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
